File: DictationAssistant/WaitingTimeAccordingToExpression.vb
Imports System.Text.RegularExpressions
Imports NCalc

Public Class WaitingTimeAccordingToExpression
    Implements IWaitingTimeCalculator
    Private ExpressionObject As Expression
    Public Sub New(ExpressionText As String)
        ExpressionObject = New Expression(ExpressionText, EvaluateOptions.IgnoreCase)
        AddHandler ExpressionObject.EvaluateFunction,
            Sub(name As String, args As FunctionArgs)
                Select Case name.ToLower()
                    Case "Length".ToLower()
                        args.Result = args.Parameters(0).Evaluate().ToString().Length
                    Case "WordCount".ToLower()
                        args.Result = WaitingTimeHelper.GetWordCount(args.Parameters(0).Evaluate().ToString())
                    Case "RegexReplace".ToLower()
                        'RegexReplace(Input, Pattern, Replacement, [IgnoreCase = False])
                        args.Result = Regex.Replace(args.Parameters(0).Evaluate().ToString(),
                                                    args.Parameters(1).Evaluate().ToString(),
                                                    args.Parameters(2).Evaluate().ToString(),
                                                    If(args.Parameters.Length >= 4 AndAlso Convert.ToBoolean(args.Parameters(3).Evaluate()), RegexOptions.IgnoreCase, RegexOptions.None)
                                                    )

                    Case "RegexMatch".ToLower()
                        'RegexMatch(Input, Pattern, [IgnoreCase = False])
                        args.Result = Regex.IsMatch(args.Parameters(0).Evaluate().ToString(),
                                                    args.Parameters(1).Evaluate().ToString(),
                                                    If(args.Parameters.Length >= 3 AndAlso Convert.ToBoolean(args.Parameters(2).Evaluate()), RegexOptions.IgnoreCase, RegexOptions.None)
                                                    )
                End Select
            End Sub
    End Sub
    Public Function CalculateWaitingTime(Word As String) As Int32 Implements IWaitingTimeCalculator.CalculateWaitingTime
        Dim EvaluateParameterCallback As EvaluateParameterHandler =
            Sub(name As String, args As ParameterArgs)
                Select Case name.ToLower()
                    Case "Length".ToLower()
                        args.Result = Word.Length
                    Case "WordCount".ToLower()
                        args.Result = WaitingTimeHelper.GetWordCount(Word)
                    Case "Text".ToLower()
                        args.Result = Word
                End Select
            End Sub

        AddHandler ExpressionObject.EvaluateParameter, EvaluateParameterCallback
        Dim ReturnValue As Int32 = Convert.ToInt32(ExpressionObject.Evaluate())
        RemoveHandler ExpressionObject.EvaluateParameter, EvaluateParameterCallback
        Return ReturnValue
    End Function
End Class

File: DictationAssistant/WaitingTimeAccordingToExpression.cs
using System;
using System.Text.RegularExpressions;
using NCalc;

namespace DictationAssistant
{
    public class WaitingTimeAccordingToExpression : IWaitingTimeCalculator
    {
        private Expression ExpressionObject;
        public WaitingTimeAccordingToExpression(string expressionText)
        {
            ExpressionObject = new Expression(expressionText, EvaluateOptions.IgnoreCase);
            ExpressionObject.EvaluateFunction += (string name, FunctionArgs args) =>
            {
                switch (name.ToLower())
                {
                    case "length":
                        {
                            args.Result = args.Parameters[0].Evaluate().ToString().Length;
                            break;
                        }

                    case "wordcount":
                        {
                            args.Result = WaitingTimeHelper.GetWordCount(args.Parameters[0].Evaluate().ToString());
                            break;
                        }

                    case "regexreplace":
                        {
                            // RegexReplace(Input, Pattern, Replacement, [IgnoreCase = False])
                            args.Result = Regex.Replace(args.Parameters[0].Evaluate().ToString(), args.Parameters[1].Evaluate().ToString(), args.Parameters[2].Evaluate().ToString(), args.Parameters.Length >= 4 && Convert.ToBoolean(args.Parameters[3].Evaluate()) ? RegexOptions.IgnoreCase : RegexOptions.None
    );
                            break;
                        }

                    case "regexmatch":
                        {
                            // RegexMatch(Input, Pattern, [IgnoreCase = False])
                            args.Result = Regex.IsMatch(args.Parameters[0].Evaluate().ToString(), args.Parameters[1].Evaluate().ToString(), args.Parameters.Length >= 3 && Convert.ToBoolean(args.Parameters[2].Evaluate()) ? RegexOptions.IgnoreCase : RegexOptions.None
    );
                            break;
                        }
                }
            };
        }
        public Int32 CalculateWaitingTime(string word)
        {
            void EvaluateParameterCallback(string name, ParameterArgs args)
            {
                switch (name.ToLower())
                {
                    case "length":
                        {
                            args.Result = word.Length;
                            break;
                        }

                    case "wordcount":
                        {
                            args.Result = WaitingTimeHelper.GetWordCount(word);
                            break;
                        }

                    case "text":
                        {
                            args.Result = word;
                            break;
                        }
                }
            }

            ExpressionObject.EvaluateParameter += EvaluateParameterCallback;
            Int32 ReturnValue = Convert.ToInt32(ExpressionObject.Evaluate());
            ExpressionObject.EvaluateParameter -= EvaluateParameterCallback;
            return ReturnValue;
        }
    }
}
